test/k8sT/Policies.go

```diff
@@ -719,6 +726,119 @@ var _ = Describe("K8sPolicyTest", func() {
 			})
 		})
 
+		Context("Redirects traffic to proxy when no policy is applied with proxy-visibility annotation", func() {
+
+			var (
+				// track which app1 pod we care about, and its corresponding
+				// cilium pod.
+				app1Pod   string
+				ciliumPod string
+				nodeName  string
+				//app1EpID        int64
+				monitorFileName = "monitor-%s.log"
+				appPods         map[string]string
+				app1PodIP       string
+			)
+
+			BeforeAll(func() {
+				appPods = helpers.GetAppPods(apps, namespaceForTest, kubectl, "id")
+				podsNodes, err := kubectl.GetPodsNodes(namespaceForTest, "-l id=app1")
+				Expect(err).To(BeNil(), "error getting pod->node mapping")
+				Expect(len(podsNodes)).To(Equal(2))
+				// Just grab the first one.
+				for k, v := range podsNodes {
+					app1Pod = k
+					nodeName = v
+					break
+				}
+
+				Expect(kubectl.WaitforPods("foo", "-l zgroup=testapp", helpers.HelperTimeout)).To(BeNil())
+				var podList v1.PodList
+				err = kubectl.GetPods(namespaceForTest, fmt.Sprintf("-n %s -l k8s-app=cilium --field-selector spec.nodeName=%s", helpers.KubeSystemNamespace, nodeName)).Unmarshal(&podList)
+				Expect(err).To(BeNil())
+
+				var app1PodModel v1.Pod
+				Expect(kubectl.Exec(fmt.Sprintf("%s get pod -n %s %s -o json", helpers.KubectlCmd, namespaceForTest, app1Pod)).Unmarshal(&app1PodModel)).To(BeNil())
+				Expect(app1PodModel).ToNot(BeNil())
+				Expect(len(podList.Items)).To(Equal(1))
+				ciliumPod = podList.Items[0].Name
+				app1PodIP = app1PodModel.Status.PodIP
+				//var app1Ep *models.Endpoint
+				var endpoints []*models.Endpoint
+				err = kubectl.ExecPodCmd(helpers.KubeSystemNamespace, ciliumPod, "cilium endpoint list -o json").Unmarshal(&endpoints)
+				Expect(err).To(BeNil())
+				for _, ep := range endpoints {
+					if ep.Status.Networking.Addressing[0].IPV4 == app1PodIP {
+						break
+					}
+				}
+			})
+
+			BeforeEach(func() {
+
+			})
+			AfterEach(func() {
+				kubectl.Exec(fmt.Sprintf("%s annotate pod %s -n %s %s-", helpers.KubectlCmd, appPods[helpers.App1], namespaceForTest, annotation.ProxyVisibility))
+			})
+
+			checkProxyRedirection := func(app1PodIP string, redirected bool) {
+				var not = " "
+
+				if !redirected {
+					not = " not "
+				}
+
+				monitorFile := fmt.Sprintf(monitorFileName, uuid.NewUUID().String())
+
+				By("Starting monitor and generating traffic which should%s redirect to proxy", not)
+				monitorStop := kubectl.MonitorStart(helpers.KubeSystemNamespace, ciliumPod, monitorFile)
+
+				// Let the monitor get started since it is started in the background.
+				time.Sleep(2 * time.Second)
```

Comment:
I'm surprised that these kinds of sleeps are necessary for getting reliable monitor output in tests, I hope that we can drastically improve this. I'm not sure why the same was not necessary in the monitor aggregation tests, is that test just relying on some race condition?

Reply:
Yeah, this one took me a while to figure out. We are launching the monitor in the background, so there could be some lag between the launching of the monitor which causes it to not capture the flow for the `curl` call below.

Reply:
https://github.com/cilium/cilium/issues/9357